pkg/io/v2/opener.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"gocloud.dev/blob"
+	"gocloud.dev/gcerrors"
+
+	"k8s.io/test-infra/pkg/io/v2/providers"
+)
+
+// Aliases to types in the standard library
+type (
+	ReadCloser  = io.ReadCloser
+	WriteCloser = io.WriteCloser
+)
+
+// Opener has methods to retrieve a Reader or Writer for the given path
+type Opener interface {
+	Reader(ctx context.Context, path string, opts *blob.ReaderOptions) (io.ReadCloser, error)
+	Writer(ctx context.Context, path string, opts *blob.WriterOptions) (io.WriteCloser, error)
+}
+
+type opener struct {
+	gcsCredentials     []byte
+	s3Credentials      []byte
+	cachedBuckets      map[string]*blob.Bucket
+	cachedBucketsMutex sync.Mutex
+}
+
+// NewOpener returns an opener that can read GCS, S3 and local paths.
+// credentialsFile may also be empty
+// For local paths it has to be empty
+// In all other cases gocloud auto-discovery is used to detect credentials, if credentialsFile is empty.
+// For more details about the possible content of the credentialsFile see pkg/io/v2/providers.GetBucket
+func NewOpener(gcsCredentialsFile, s3CredentialsFile string) (Opener, error) {
+	var gcsCredentials []byte
+	var s3Credentials []byte
+	var err error
+	if gcsCredentialsFile != "" {
+		gcsCredentials, err = ioutil.ReadFile(gcsCredentialsFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if s3CredentialsFile != "" {
+		s3Credentials, err = ioutil.ReadFile(s3CredentialsFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &opener{
+		gcsCredentials: gcsCredentials,
+		s3Credentials:  s3Credentials,
+		cachedBuckets:  map[string]*blob.Bucket{},
+	}, nil
+}
+
+// getBucket opens a bucket
+// The storageProvider is discovered based on the given path.
+// The buckets are cached per bucket name. So we don't open a bucket multiple times in the same process
```

Comment:
I had a previous implementation which returned a readerCloser which contained the bucket and the reader/writer. The close method both closed the reader and the writer. To simplify this we decided to just return the reader and writer and cache the bucket so it doesn't have to be closed.

I see multiple options:
* return a readerCloser / writerCloser again which encapsulates the bucket and the reader/writer and closes both: See: https://github.com/kubernetes/test-infra/commit/f75838e0ee47004e4faedc0099e16f93871aafea#diff-12f70e7ec70b1a60275ef5d6dab027beR101
* keep it as is
* only return reader/ writer and ignore the bucket.Close(). (That might be an option because all implemented providers in gophercloud currently have a Close() implementation like this:
   ```
      func (b *bucket) Close() error {
	return nil
      }
   ```